modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts

```diff
@@ -100,14 +99,38 @@ export async function prove(p: bigint, q: bigint): Promise<DeserializedPaillierB
   // This is calculating the inverse of the function y^4 mod N,
   // i.e.y ^ (1 / 4), where N = pq is a blum integer using HOC - Fact 2.160
   // from cacr.uwaterloo.ca / hac / about / chap2.pdf
-  // Prover generates y_i.
+  // Prover generates a_i, b_i, and y_i.
   const y = generateY(N, w);
   // Prover calculates z_i = y_i ^ d mod N
   const z = y.map((y_i) => modPow(y_i, d, N));
   // Prover calculates x_i = y_i ^ 1/4 mod N using [HOC - Fact 2.160]
   const e = ((l + BigInt(4)) / BigInt(8)) ** BigInt(2);
-  const x = y.map((y_i) => modPow(y_i, e, N));
-  return { w, x, z };
+  const ab = Array(m).fill(null);
+  const x = await Promise.all(
+    y.map(async (y_i, i) => {
+      // Select random a_i, b_i so that y_i' = (-1)^{a_i} * w^{b_i} * y is
+      // quadratic residue of N using [HOC - Fact 2.137].
+      let t;
+      while (true) {
+        t = y_i;
+        ab[i] = (await randBits(2))[0];
+        if (ab[i] & 2) {
+          t = -t;
+        }
+        if (ab[i] & 1) {
+          t *= w;
+        }
+        if (
+          modPow(t, (p - BigInt(1)) / BigInt(2), p) === BigInt(1) &&
+          modPow(t, (q - BigInt(1)) / BigInt(2), q) === BigInt(1)
```

Comment:
This can also be done by checking if y_i is quadratic residue of p or q first by calculating
modPow(y_i, (p - 1) / 2, p), modPow(t, (q - 1) / 2, q), modPow(w, (p -1)/2, p) and modPow(w, (q -1)/2, q) as above, the result can be used to determine the value of a_i and b_i.  This does need 2 more calculations, so on average it evens out with the try-and-error approach.

Since p, q === 3 mod 4 and are primes, "If p ≡ 3 (mod 4) the negative of a residue modulo p is a nonresidue and the negative of a nonresidue is a residue." according to https://en.wikipedia.org/wiki/Quadratic_residue.  Jacobi(w, N) = -1 means w is quadratic residue to one of p, q and nonresidue to the other.

For example, if (y_i/p) = 1 (residue), (y_i/q) = -1 (nonresidue), (w/p) = 1, (w/q) = -1, then a_i = 0, b_i = 1.